libraries/AP_HAL_ChibiOS/Storage.cpp

```diff
@@ -126,12 +126,69 @@ void Storage::_save_backup(void)
 {
 #ifdef USE_POSIX
     // allow for fallback to microSD based storage
-    if (sdcard_retry()) {
-        int fd = AP::FS().open(HAL_STORAGE_BACKUP_FILE, O_WRONLY|O_CREAT|O_TRUNC);
-        if (fd != -1) {
-            AP::FS().write(fd, _buffer, CH_STORAGE_SIZE);
-            AP::FS().close(fd);
+    // create the backup directory if need be
+    int ret;
+    const char* _storage_bak_directory = HAL_STORAGE_BACKUP_FOLDER;
+
+    EXPECT_DELAY_MS(3000);
+
+    // We want to do this desperately,
+    // So we keep trying this for a second
+    uint32_t start_millis = AP_HAL::millis();
+    while(!sdcard_retry() && (AP_HAL::millis() - start_millis) < 1000) {
+        hal.scheduler->delay(1);        
+    }
+
+    ret = AP::FS().mkdir(_storage_bak_directory);
+    if (ret == -1 && errno != EEXIST) {
+        return;
+    }
+
+    char* fname = nullptr;
+    unsigned curr_bak = 0;
+    ret = asprintf(&fname, "%s/last_storage_bak", _storage_bak_directory);
+    if (fname == nullptr && (ret <= 0)) {
+        return;
+    }
+    int fd = AP::FS().open(fname, O_RDONLY);
+    if (fd != -1) {
+        char buf[10];
+        memset(buf, 0, sizeof(buf));
+        if (AP::FS().read(fd, buf, sizeof(buf)-1) > 0) {
+            //only record last 100 backups
+            curr_bak = (strtol(buf, NULL, 10) + 1)%100;
```

Comment:
instead of hard-coding 100, add a HAL_STORAGE_BACKUP_COUNT, with #ifndef/#define